sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastToCpuExec.scala

```diff
@@ -83,36 +87,27 @@ case class GpuBroadcastToCpuExec(override val mode: BroadcastMode, child: SparkP
             }
 
             // deserialize to host buffers in the driver and then convert to rows
-            val dataTypes = child.output.map(_.dataType)
-
-            val gpuBatches = serializedBatches.safeMap { cb =>
-              val hostColumns = (0 until cb.header.getNumColumns).map { i =>
-                val columnHeader = cb.header.getColumnHeader(i)
-                val hcv = new HostColumnVector(
-                  columnHeader.getType,
-                  columnHeader.rowCount,
-                  Optional.of(columnHeader.nullCount),
-                  cb.buffer,
-                  null, null, List.empty.asJava)
-                new RapidsHostColumnVector(dataTypes(i), hcv)
+            val gpuBatches = withResource(serializedBatches) { _ =>
+              serializedBatches.safeMap { cb =>
+                val hostColumns = GpuBroadcastToCpuExec.buildHostColumns(cb.header, cb.buffer,
+                  dataTypes)
+                val rowCount = cb.header.getNumRows
+                new ColumnarBatch(hostColumns.toArray, rowCount)
               }
-              val rowCount = hostColumns.headOption.map(_.getRowCount.toInt).getOrElse(0)
-              new ColumnarBatch(hostColumns.toArray, rowCount)
             }
 
             val broadcasted = try {
-              val rows = new ListBuffer[InternalRow]()
-              gpuBatches.foreach(cb => rows.appendAll(cb.rowIterator().asScala))
-
-              val numRows = rows.length
+              val numRows = gpuBatches.map(_.numRows).sum
               checkRowLimit(numRows)
               numOutputRows += numRows
 
               val relation = withResource(new NvtxWithMetrics(
                 "broadcast build", NvtxColor.DARK_GREEN, buildTime)) { _ =>
                 val toUnsafe = UnsafeProjection.create(output, output)
-                val unsafeRows = rows.iterator.map(toUnsafe)
-                 val relation = ShimLoader.getSparkShims
+                val unsafeRows = gpuBatches.flatMap {
+                  _.rowIterator().asScala.map(r => toUnsafe(r).copy())
```

Comment:
nit, it may be worth adding a comment above this explaining the reason for the copy. I think folks can backtrack to this github issue, and it can be done later too, not to block this PR from an expensive CI run today.